### DISCUSSION TITLE
Welcome to ocpp Discussions!

### DISCUSSION THREAD

**Discussion by OrangeTux** [mobilityhouse/ocpp#163] [General]
## 👋 Welcome!
  We’re using Discussions as a place to connect with other members of our community. We hope that you:
  * Ask questions you’re wondering about.
  * Share ideas.
  * Engage with other community members.
  * Welcome others and are open-minded. Remember that this is a community we
  build together 💪.

  To get started, comment below with an introduction of yourself and tell us about what you do with this community.

**Comment by benoit-bremaud** [comment]
Hello everyone,

I'm Benoit, a junior web developer based in France, currently engaged in an exciting project within an IT company that's pioneering in the realm of electric vehicle (EV) charging station fleet management. My endeavor involves developing a cutting-edge web application designed to assess the compliance of charging stations with the OCPP 1.6 and OCPP 2.0 protocols.

This project is a significant part of my journey in the tech industry, where I am harnessing the power of Python to create a solution that stands at the intersection of technology and sustainability. My focus is on utilizing Python's versatility and robust ecosystem to build an application that not only simplifies the compliance verification process but also contributes to the reliability and efficiency of EV charging infrastructure.

Being deeply committed to this project, I aim to leverage my skills in Python to tackle the challenges it presents and to contribute meaningful advancements in the EV sector. I'm keen to share my experiences, delve into technical discussions, and learn from the diverse perspectives within this community.

I look forward to engaging with fellow members, exchanging knowledge, and collaborating on ideas that push the envelope of what's possible in our field.

Let's embark on this journey together and drive positive change through technology. 💪

**Comment by azamau** [comment]
Hello Team,
I am starting new with OCPP, I want to use this python package and implement it, can some one please tell me instruction on how to run this package. 

Thanks

**Reply by OrangeTux** [reply]
Except from the [examples](https://github.com/mobilityhouse/ocpp/tree/master/examples) , you can't "run" this package. This library provides the building blocks to build your own programs that you can run.

**Reply by azamau** [reply]
Thank you for your prompt response. My intention is to utilize this application as a comprehensive tool.

I am interested in developing an OCPP Central Management System (CMS) specifically designed for testing chargers. Could someone please provide assistance in this endeavor?

Your support in this matter would be greatly appreciated.

**Comment by Nwangyh** [comment]
![image](https://github.com/mobilityhouse/ocpp/assets/161010895/b76251f4-2a76-4f19-a035-c115a9fd3879)
As far as I know, ocpp 2.0.1 released the latest edition3 version on the official website, and provided many test cases similar to OCTT. In the testcases file, can you try to make a public OCTT test program?

**Comment by MaciejGGH** [comment]
Hi!

My goal is to create a server that I could eventualy use to control my own home charger. I have some Python experience, but I'm not a seasoned dev. For start I'm just trying to run the examples. Once this will be running, I'll try to make my charger talk to central_system.py.

I'm running the `central_system.py`:

    PS C:\Projects\ocpp> & c:/Projects/ocpp/.venv/Scripts/python.exe C:\Projects\ocpp\central_system.py
    INFO:websockets.server:server listening on 0.0.0.0:9000
    INFO:root:Server Started listening to new connections...

I got this notification: `C:\Projects\ocpp\.venv\Lib\site-packages\ocpp\v201\call.py:511: UserWarning: BootNotificationPayload is deprecated, use instead BootNotification`, so I've changed the call to:

        request = call.BootNotification(
            charging_station={"model": "Wallbox XYZ", "vendor_name": "anewone"},
            reason="PowerUp",
        ) 

However, the `charge_point.py` throws:

    PS C:\Projects\ocpp> & c:/Projects/ocpp/.venv/Scripts/python.exe c:/Projects/ocpp/charge_point.py
    INFO:ocpp:CP_1: send [2,"94cfbd48-c48d-4990-ac41-c530ec37bfea","BootNotification",{"chargingStation":{"model":"Wallbox XYZ","vendorName":"anewone"},"reason":"PowerUp"}]
    Traceback (most recent call last):
      File "c:\Projects\ocpp\charge_point.py", line 58, in <module>
        asyncio.run(main())
        ~~~~~~~~~~~^^^^^^^^
      File "C:\Users\Maciej\AppData\Local\Programs\Python\Python313\Lib\asyncio\runners.py", line 194, in run
        return runner.run(main)
               ~~~~~~~~~~^^^^^^
      File "C:\Users\Maciej\AppData\Local\Programs\Python\Python313\Lib\asyncio\runners.py", line 118, in run
        return self._loop.run_until_complete(task)
               ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~^^^^^^
      File "C:\Users\Maciej\AppData\Local\Programs\Python\Python313\Lib\asyncio\base_events.py", line 721, in run_until_complete
        return future.result()
               ~~~~~~~~~~~~~^^
      File "c:\Projects\ocpp\charge_point.py", line 51, in main
        await asyncio.gather(
            charge_point.start(), charge_point.send_boot_notification()
        )
      File "C:\Projects\ocpp\.venv\Lib\site-packages\ocpp\charge_point.py", line 239, in start
        message = await self._connection.recv()
                  ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
      File "C:\Projects\ocpp\.venv\Lib\site-packages\websockets\asyncio\connection.py", line 313, in recv
        raise self.protocol.close_exc from self.recv_exc
    websockets.exceptions.ConnectionClosedError: received 1011 (internal error); then sent 1011 (internal error)

In the console for `central_system.py` I get:

    ERROR:websockets.server:connection handler failed
    Traceback (most recent call last):
      File "C:\Projects\ocpp\.venv\Lib\site-packages\websockets\asyncio\server.py", line 373, in conn_handler
        await self.handler(connection)
              ~~~~~~~~~~~~^^^^^^^^^^^^
    TypeError: on_connect() missing 1 required positional argument: 'path'

I also tried to set default value of None for path (`async def on_connect(websocket, path=None):`), then I get a different error thrown by `charge_point.py`:

    ERROR:websockets.server:connection handler failed
    Traceback (most recent call last):
      File "C:\Projects\ocpp\.venv\Lib\site-packages\websockets\asyncio\server.py", line 373, in conn_handler
        await self.handler(connection)
      File "C:\Projects\ocpp\central_system.py", line 43, in on_connect
        requested_protocols = websocket.request_headers["Sec-WebSocket-Protocol"]
                              ^^^^^^^^^^^^^^^^^^^^^^^^^
    AttributeError: 'ServerConnection' object has no attribute 'request_headers'


Please advise.

**Reply by Paulerpro** [reply]
TypeError: on_connect() missing 1 required positional argument: 'path'

if you get this error, kindly downgrade your the version of your websocket library to 10.

run pip install websockets==10

this helped me

**Reply by MaciejGGH** [reply]
That was it, many thanks!

**Comment by Paulerpro** [comment]
Sorry no one could respond to your question at the time. How did the OCPP
project play out, Maciej?

**Reply by MaciejGGH** [reply]
It didn't... yet! Stuck on the above problem I kept myself busy with other stuff. Now that I have the example working, I just added small script to check hourly energy prices at my local provider. Now I will work on the basics a bit more, try to hook my charger to my Very Own  Central System and try to make it work :)

**Comment by Paulerpro** [comment]
Happy to hear you are back to business. Congrats.

Oh you mentioned you’re ain’t a seasoned dev, but it takes quite a handful
experience in python to take on such a solid project. lol

Do you plan to go commercial with your setup? Maybe MVP.

**Reply by MaciejGGH** [reply]
I can be persistent :) I don't have big plans, but who knows? First I want to make it work. Then I'll thing of making it available. Eventually I might think of going further, we'll see.

**Comment by tannyle289** [comment]
Hi, first of all thank you so much for providing such an amazing tool!! Keep it up the good pace 🙏🙏

I'm using the `ocpp` library to generate synthetic OCPP 1.6 message sequences for training ML models.
I want to make sure the generated data reflects real charger behavior. A few questions for anyone who has worked with real OCPP deployments 🙏:

- `MeterValues` interval — Is 60 seconds the most common `MeterValueSampleInterval` in practice, or do most operators configure something different (30s, 120s, 300s)?
- Which `measurands` do real chargers actually report? The spec lists many (`Voltage`, `Current.Import`, `Energy.Active.Import.Register`, `SoC`, `Temperature`, `Power.Active.Import`, etc.), but I suspect most AC chargers only report a subset. Which ones are typically available on a standard 7-22kW AC charger or other common ones?
  - Ref: https://ocpp.md/ocpp-1.6j/schemas/core/#:~:text=%23-,MeterValues,-Direction%3A%20CP
- `Heartbeat interval` — Is 300 seconds standard, or do operators typically configure a different value?
- `StatusNotification` error codes — In practice, which `ChargePointErrorCode` values appear most frequently on real chargers? Are some (like `HighTemperature`, `GroundFailure`) very rare while others (like `InternalError`) are common?
- Session patterns — For a public charger in an urban setting, is 3-5 sessions per day a reasonable assumption? And typical session duration of 30-60 minutes for AC charging?

Any insights from real deployments would be very helpful for making my synthetic dataset more realistic. Happy to share the generator once it's ready — it produces full OCPP 1.6J and OCPP 2.1 message sequences with configurable degradation patterns.

Thanks a lot 🙏!

**Reply by Nwangyh** [reply]
Hello!
Regarding the first question: the upload is actually performed according to the monitoring interval required by the cloud operation platform. Generally, 30 seconds and 60 seconds are the most common options, but the specific choice also depends on the charging power.
Under normal circumstances, 60 seconds is sufficient for low‑power AC/DC charging. For high‑power charging (e.g., 22 kW and above), a shorter MeterValueSampleInterval may be a better choice.
Regarding the second question: the commonly reported information mainly includes typical parameters such as output current, output voltage, (electronic lock status for DC charging piles), insulation detection, and other fault statuses (mostly custom-defined).
The Heartbeat interval is not a standard fixed value. If the network condition is good, the heartbeat can even be set to 12 hours (43,200 seconds). This requires confidence that the connection will not drop, and that the clock deviation of your hardware device will not be significant within 12 hours.
For error codes, you can refer to common categories:
Electrical faults: voltage, current, insulation detection for DC piles, electronic lock status, etc.
Environmental faults: ambient temperature, pile body temperature, DC power module temperature, PE short circuit, ground disconnection, short circuit to ground, output short circuit, etc. — depending on which faults the charging pile controller can detect.
For public charging piles in urban environments:
Low‑power piles usually have about 3–5 sessions per day.
High‑power DC charging piles of 22 kW typically have 10 or more sessions per day, with an average charging time of 1–2 hours per vehicle (nominal 1.5 hours).

For reference only. Hope this helps!

**Reply by tannyle289** [reply]
Hello again,

Wow!! Thank you so much for the detailed reply! This is extremely helpful and confirms many of my assumptions 🙏🙏

A few follow-up questions based on your experience, if you don't mind:

- I found `ClockAlignedDataInterval` in the [OCPP 1.6 Core Profile config keys](https://ocpp.md/ocpp-1.6j/#:~:text=%23-,6.1%20Core%20Profile,-KEY). From my reading, it seems like it's basically `MeterValueSampleInterval` but it can report even when no car is plugged in — is that correct?
   
     - In real deployments, is it almost always set to 0 (disabled)? (Not sure if I should add it to the synthetize process)
     - If enabled, the charger report the `measurands` during idle time (no car charging) with configurable values via `MeterValuesAlignedData` after each `ClockAlignedDataInterval `?

- Lastly the practical question, let say when a charger has issues and a technician is dispatched:

    - Do they primarily rely on OCPP error codes (StatusNotification) and logs; or do they perform separate diagnostic tools / physical measurements (multimeter, thermal camera, etc.) that is irrelevant to OCPP?

Thanks again 🙏!

**Reply by abhasdudeja** [reply]
Hi, so.
ClockAlignedDataInterval is used in India to sync meter values both Input and Output with the Central System and as you rightly said, sends data even when there is no vehicle plugged in and it is almost similar to MeterValueSampledInterval. ClockAligned sends meter values with transaction id as 0. 

In real deployments, it depends on the use case, I have seen both 0 and 900 as the value.
When enabled, it should ideally use MeterValueAlignedData to send measurands.

About your third question, i have observed engineers to first request OCPP logs, then diagnostic logs, and then they go around doing anything physical with the charger.

**Reply by tannyle289** [reply]
Hi again,

Thank you so much for your answers 🙏!

Another follow up question popped up about the technician dispatch
- Since you mentioned diagnostic logs, are they another kind of logs independent from OCPP logs, and do they follow specific format, or protocol like OCPP?

Thanks again!

**Reply by ghost** [reply]
Diagnostic Logs are taken from the charger directly (Get Diagnostic Logs - Command), they are independent of the OCPP logs. They should ideally have a format or standard, but it is entirely OEM dependent.
They contain Module, Controller and Communication level logs of the charger. They are in much more depth than the OCPP ones. Often includes communication flow between the Charger and Vehicle.
